### PR TITLE
[metal] Support `KernelReturnStmt`

### PR DESCRIPTION
**Pull request by k-ye**
This is probably mentioned somewhere, but I think we should add a new `rets` in `Context` as well?

https://github.com/taichi-dev/taichi/blob/bbddaa10def6e2e9e566ec8f665daf480aab652e/taichi/runtime/llvm/context.h#L15-L18

Once we do that, OpenGL/Metal can copy the return values back to `Context::rets`, so that we don't have to overuse `Context::args` for holding return values:

https://github.com/taichi-dev/taichi/blob/bbddaa10def6e2e9e566ec8f665daf480aab652e/taichi/program/program.cpp#L500

Related issue = #909 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
Actually, I feel like maybe we should make `Context` read-only (which is required by the CUDA backend when running without unified memory), then we use a separate buffer for the return values. Or we should just store a pointer in `Context` for the return value buffer. (Of course, I have no idea if this is the best way to do it for Metal/OpenGL)

File: taichi/backends/metal/codegen_metal.cpp [hunk 1]
```diff
@@ -270,6 +270,11 @@ class KernelCodegen : public IRVisitor {
          stmt->val->raw_name());
```
Comment (archibate) [inline taichi/backends/metal/codegen_metal.cpp:270]: We'll one day make another PR to remove `ArgStore` and `mark_as_return_value` completely.

File: taichi/backends/metal/kernel_util.h [hunk 2]
```diff
@@ -63,48 +63,82 @@ struct KernelAttributes {
...
+// the return values of a Taichi kernel.
+//
 // Note that all Metal kernels belonging to the same Taichi kernel will share
-// the same kernel args (attributes + Metal buffer). This is because kernel
-// arguments is a Taichi-level concept.
+// the same kernel args (i.e. they use the same Metal buffer for input args and
+// return values). This is because kernel arguments is a Taichi-level concept.
+//
+// TODO(#909): Rename to KernelArgsRetsAttributes
 class KernelArgsAttributes {
```
Comment (archibate) [inline taichi/backends/metal/kernel_util.h:74]: Why not apply in this PR?

File: taichi/backends/metal/kernel_util.h [hunk 3]
```diff
@@ -63,48 +63,82 @@ struct KernelAttributes {
...
+  }
+
+  // Returns true if the kernel has neither input args nor return values.
+  inline bool empty() const {
+    return !(has_args() || has_rets());
+  }
+
+  // Total size in bytes of the input args and return values
+  inline size_t args_rets_bytes() const {
+    return args_rets_bytes_;
```
Comment (archibate) [inline taichi/backends/metal/kernel_util.h:120]: nit: args_rets_bytes sounds to complicated, what about io_bytes, or ctx_bytes?

File: taichi/backends/metal/kernel_util.h (hunk 2 above)
Comment (archibate) [inline taichi/backends/metal/kernel_util.h:74]:
```suggestion
class KerneCtxAttributes {
```

File: taichi/backends/metal/kernel_util.h [hunk 4]
```diff
@@ -63,48 +63,82 @@ struct KernelAttributes {
...
-    bool is_return_val = false;
   };
 
-  explicit KernelArgsAttributes(const std::vector<Kernel::Arg> &args);
+ public:
+  // This is mostly the same as Kernel::Arg, with Metal specific attributes.
+  struct ArgAttributes : public AttribsBase {};
+
+  // This is mostly the same as Kernel::Ret, with Metal specific attributes.
+  struct RetAttributes : public AttribsBase {};
```
Comment (archibate) [inline taichi/backends/metal/kernel_util.h:93]:
```suggestion
  struct OutputAttributes : public AttribsBase {};
```

File: taichi/backends/metal/kernel_util.h [hunk 5]
```diff
@@ -63,48 +63,82 @@ struct KernelAttributes {
...
     int index = -1;
     MetalDataType dt;
     bool is_array = false;
-    bool is_return_val = false;
   };
 
-  explicit KernelArgsAttributes(const std::vector<Kernel::Arg> &args);
+ public:
+  // This is mostly the same as Kernel::Arg, with Metal specific attributes.
+  struct ArgAttributes : public AttribsBase {};
```
Comment (archibate) [inline taichi/backends/metal/kernel_util.h:90]:
```suggestion
  struct InputAttributes : public AttribsBase {};
```

File: taichi/backends/metal/codegen_metal.cpp [hunk 6]
```diff
@@ -80,7 +80,7 @@ class KernelCodegen : public IRVisitor {
         kernel_(kernel),
         compiled_structs_(compiled_structs),
         needs_root_buffer_(compiled_structs_->root_size > 0),
-        args_attribs_(kernel_->args) {
+        args_attribs_(*kernel_) {
```
Comment (archibate) [inline taichi/backends/metal/codegen_metal.cpp:83]:
```suggestion
        ctx_attribs_(*kernel_) {
```

**Comment by k-ye** [comment]
@archibate 

Thanks for all the nits, especially the suggested changes! However, note that some of the name changes need to be applied to more than one place, so it's better if you could just tell me what to change, and I will change all of them locally :)

---

I thought about calling these Input/Output Args, instead of Input Args/Return Values. These way the name is unified to "Args",  but then I realized a nuance here:

1. Return value means the value returned by the kernel, obviously.
2. But we *do* have Output Args, even at this point. The numpy arrays returned by `to_numpy()` is such an example. It is passed in to the kernel, but its data also needs to be copied back to the host.

So I'd like to stay with Arg and Ret, which is also consistent with the current naming in `Kernel` created by you.

File: taichi/backends/metal/kernel_util.h (hunk 3 above)
Comment (k-ye) [reply taichi/backends/metal/kernel_util.h:120]: Maybe just ctx_bytes, since you've suggested to name the class `KernelCtxAttributes`?

File: taichi/backends/metal/codegen_metal.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/backends/metal/codegen_metal.cpp:270]: Yep, sounds good. FWIW, I don't think `ArgStoreStmt` is in use anymore, so maybe it's fine to just remove it after this one?

File: taichi/backends/metal/kernel_util.h (hunk 2 above)
Comment (k-ye) [reply taichi/backends/metal/kernel_util.h:74]: I thought the renaming here might make it harder to review. But I'll do it since you don't object..

Comment (k-ye) [reply taichi/backends/metal/kernel_util.h]: Done

File: taichi/backends/metal/kernel_util.h (hunk 3 above)
Comment (k-ye) [reply taichi/backends/metal/kernel_util.h]: Done

File: taichi/backends/metal/kernel_util.h (hunk 2 above)
Comment (k-ye) [reply taichi/backends/metal/kernel_util.h]: Done

File: taichi/backends/metal/kernel_util.h (hunk 4 above)
Comment (k-ye) [reply taichi/backends/metal/kernel_util.h:91]: FYI, I'd like to keep `Args` and `Ret`, and explained my opinion below. This way it is also consistent with `Kernel::Arg` and `Kernel::Ret`.

File: taichi/backends/metal/codegen_metal.cpp (hunk 6 above)
Comment (k-ye) [reply taichi/backends/metal/codegen_metal.cpp]: Done

**Comment by archibate** [comment]
> Actually, I feel like maybe we should make Context read-only (which is required by the CUDA backend when running without unified memory), then we use a separate buffer for the return values. Or we should just store a pointer in Context for the return value buffer. (Of course, I have no idea if this is the best way to do it for Metal/OpenGL)

Thank for the info. fyi now metal concats args and rets buffer, args + rets. opengl shares args and rets buffer, max(args, rets).

**Comment by k-ye** [comment]
> fyi now metal concats args and rets buffer, args + rets. opengl shares args and rets buffer, max(args, rets).

Yep, I remember that you also plan to do similar concatenation eventually?
